delta/plugins/project-deletion/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/projectdeletion/ProjectDeleter.scala

```diff
@@ -0,0 +1,99 @@
+package ch.epfl.bluebrain.nexus.delta.plugins.projectdeletion
+
+import cats.syntax.all._
+import ch.epfl.bluebrain.nexus.delta.kernel.syntax.instantSyntax
+import ch.epfl.bluebrain.nexus.delta.kernel.utils.IOUtils
+import ch.epfl.bluebrain.nexus.delta.plugins.projectdeletion.model.ProjectDeletionConfig
+import ch.epfl.bluebrain.nexus.delta.sdk.ProjectResource
+import monix.bio.UIO
+import ProjectDeleter._
+
+import java.time.{Duration, Instant}
+
+object ProjectDeleter {
+
+  def projectDeletionPass(
+      allProjects: UIO[Seq[ProjectResource]],
+      deleteProject: ProjectResource => UIO[Unit],
+      config: ProjectDeletionConfig,
+      lastEventTime: (ProjectResource, Instant) => UIO[Instant]
+  ): UIO[Unit] = {
+
+    val deleter = new ProjectDeleter(deleteProject, config, lastEventTime)
+
+    for {
+      allProjects <- allProjects
+      now         <- IOUtils.instant
+      _           <- allProjects.traverse(deleter.processProject(_, now))
+    } yield {
+      ()
+    }
+  }
+
+  implicit class BooleanTaskOps(val left: Boolean) extends AnyVal {
+    def `<||>`(right: UIO[Boolean]): UIO[Boolean] = {
+      if (left) {
+        UIO.pure(true)
+      } else {
+        right
+      }
+    }
+
+    def `<&&>`(right: UIO[Boolean]): UIO[Boolean] = {
+      if (left) {
+        right
+      } else {
+        UIO.pure(false)
+      }
+    }
+  }
+}
+
+class ProjectDeleter(
```

Comment:
We don't really need a class here, everything could be handled in the object

Reply:
I think the encapsulation here provides value

Reply:
It looks less consistent that way as it is not the same syntax if you want to process one project or a list of projects

Reply:
Isn't the difference just the new keyword?

Reply:
In the second case, the parameters are scattered and you also need to provide the now